### PR TITLE
CBG-455: Handle inconsistent tombstone format for deltas vs. non-deltas

### PR DESCRIPTION
**Pull request by bbrks**

### CODE REVIEW

File: rest/blip_api_test.go [hunk 1]
```diff
@@ -1702,6 +1702,96 @@ func TestBlipDeltaSyncPullRemoved(t *testing.T) {
...
+
+	msg, ok := client.pullReplication.WaitForMessage(5)
+	assert.True(t, ok)
+	msgBody, err := msg.Body()
+	assert.NoError(t, err)
+	assert.Equal(t, `{}`, string(msgBody))
+	assert.Equal(t, "1", msg.Properties[revMessageDeleted])
+}
+
+// TestBlipDeltaSyncPullTombstonedStarChan tests a simple pull replication that deletes a document in the star channel.
```
Comment (bbrks) [inline rest/blip_api_test.go]:
```suggestion
// TestBlipDeltaSyncPullTombstonedStarChan tests a simple pull replication that deletes a document when the user has access to the star channel.
```

File: db/revision_cache_interface.go [hunk 2]
```diff
@@ -101,15 +101,17 @@ type RevisionDelta struct {
 	AttachmentDigests []string // Digests for all attachments present on ToRevID
 	ToChannels        base.Set // Full list of channels for the to revision
 	RevisionHistory   []string // Revision history from parent of ToRevID to source revID, in descending order
+	ToDeleted         bool
```
Comment (bbrks) [inline db/revision_cache_interface.go]:
```suggestion
	ToDeleted         bool     // Flag if ToRevID is a tombstone
```

File: db/revision_cache_interface.go [hunk 3]
```diff
@@ -101,15 +101,17 @@ type RevisionDelta struct {
 	AttachmentDigests []string // Digests for all attachments present on ToRevID
 	ToChannels        base.Set // Full list of channels for the to revision
 	RevisionHistory   []string // Revision history from parent of ToRevID to source revID, in descending order
+	ToDeleted         bool     // Flag if ToRevID is a tombstone
 }
 
-func newRevCacheDelta(deltaBytes []byte, fromRevID string, toRevision DocumentRevision) *RevisionDelta {
+func newRevCacheDelta(deltaBytes []byte, fromRevID string, toRevision DocumentRevision, deleted bool) *RevisionDelta {
```
Comment (adamcfraser) [inline db/revision_cache_interface.go:107]: It feels like deleted should be added as a property on DocumentRevision, but we could leave that as a Mercury enhancement. What do you think about filing a tracking ticket for that?

File: db/crud.go [hunk 4]
```diff
@@ -355,11 +355,19 @@ func (db *Database) GetDelta(docID, fromRevID, toRevID string) (delta *RevisionD
...
 		if !isAuthorized {
 			return nil, redactedBody, nil
 		}
 
+		// If the revision we're generating a delta to is a tombstone, mark it as such and don't bother generating a delta
+		if deleted {
+			revCacheDelta := newRevCacheDelta([]byte(`{}`), fromRevID, toRevision, deleted)
+			db.revisionCache.UpdateDelta(docID, fromRevID, revCacheDelta)
+			return revCacheDelta, nil, nil
+		}
```
Comment (adamcfraser) [inline db/crud.go:369]: Do the unit tests exercise the case where an initial request generates the empty delta with deleted=true, and then a subsequent request retrieves that delta from the cache?  The logic looks sound, but would be good to have it validated.

**Review by adamcfraser — COMMENTED** [review]
Looks good, a couple of questions about test coverage and future enhancements.

File: db/revision_cache_interface.go (hunk 3 above)
Comment (bbrks) [reply db/revision_cache_interface.go:107]: I think Jacques tech spec covers this, but I'll check and file an issue if not.

Edit: His covers adding deleted to a Document struct. I'll include this DocumentRevision change in my revcache updates.

File: db/crud.go (hunk 4 above)
Comment (bbrks) [reply db/crud.go:369]: I should be able to get a test to force that to happen yeah. In the real world CBL shouldn't be asking for a delta from a tombstoned revision, but I'll update the test tomorrow anyway to get that covered.

Comment (adamcfraser) [reply db/crud.go:369]: I was asking about a delta *to* a tombstoned revision.  Basically the same case that TestBlipDeltaSyncPullTombstoned covers, but where the (empty) delta with deleted=true is already resident in the cache from a previous request.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**